### ISSUE TITLE
FSQ-AE 구현 및 학습 결과 (4 모델 ablation, 95.57% emotion preservation)

### ISSUE THREAD

**Issue by zxxxv** [Pachyhead/capstone_design_2026#9]
# FSQ-AE 구현 및 학습 결과 (95.57% emotion preservation)

## 개요

emotion2vec 1024d 임베딩을 FSQ Autoencoder로 압축. 4 모델 ablation 결과 **9d Skip+KL+Fine-tune 모델**이 1,568배 압축에서 **95.57% emotion 분류 보존** 달성

## 실험 구성

| Model | Architecture | Loss | Dim | Bits | 압축률 |
|---|---|---|---|---|---|
| Baseline | 3-layer MLP | MSE | 8 | 18.58 | 1,763× |
| Skip + KL (8d) | Skip + Deep MLP (4 blocks) | MSE + 0.1·KL | 8 | 18.58 | 1,763× |
| Skip + KL (9d) | Skip + Deep MLP (4 blocks) | MSE + 0.1·KL | 9 | 20.90 | 1,568× |
| + Fine-tune | (위) + lr=1e-4, 50 ep | MSE + 0.1·KL | 9 | 20.90 | 1,568× |

### Baseline
```
Encoder: 1024 0 -> 512 -> 256 -> 8 (Linear + LayerNorm + GELU)
Decoder: 거울상
파라미터: ~1.3M
```

### Skip + Deep MLP
```
MLPBlock: x + fc2(GELU(fc1(LayerNorm(x))))    # Pre-norm + Residual
          fc1: dim -> dim*2, fc2: dim*2 -> dim

Encoder: Linear(1024, 512) -> MLPBlock × 4 -> LayerNorm -> Linear(512, bottleneck)
Decoder: 거울상
파라미터: ~9.5M
```

### FSQ
- 8d: levels=[5]×8 $\rightarrow$ 5^8 = 390,625 codes
- 9d: levels=[5]×9 $\rightarrow$ 5^9 = 1,953,125 codes
- STE: `z + (z.round() - z).detach()`

## Loss

```python
# Baseline
loss = F.mse_loss(x_recon, x)

# 개선 모델
p_orig = softmax(emotion_head(x_orig))
log_p_recon = log_softmax(emotion_head(x_recon))
kl = F.kl_div(log_p_recon, p_orig, reduction="batchmean")
loss = mse + 0.1 * kl
```

KL loss가 emotion 분류 분포 보존을 직접 학습

## 데이터

- AI Hub SN=466, 17만 $\rightarrow$ 전처리 (unk 제거 + 감정별 7000 cap) $\rightarrow$ **48,148 발화**
- 화자 단위 split: train 1,144명 / val 127명 (겹침 0)
- 차원별 z-score 정규화

## 결과 1: 정량 지표

| Model | val MSE | val cos | val match |
|---|---|---|---|
| Baseline | 0.0259 | 0.9838 | - |
| Skip + KL (8d) | 0.0256 | 0.9845 | 93.72% |
| Skip + KL (9d) | 0.0199 | 0.9881 | 94.53% |
| + Fine-tune | **0.0186** | **0.9886** | **94.84%** |

## 결과 2: 감정 보존 (전체 48,148)

emotion2vec+ classifier head로 원본 vs 복원 분류 일치율 측정

### 종합

| Model | 분류 일치율 | Baseline 대비 |
|---|---|---|
| Baseline | 91.99% | reference |
| Skip + KL (8d) | 94.20% | +2.21p |
| Skip + KL (9d) | 95.16% | +3.17p |
| + Fine-tune | **95.57%** | **+3.58p** |

### 감정별 비교

| Emotion | Baseline | 8d Skip+KL | 9d Skip+KL | 9d Fine-tune |
|---|---|---|---|---|
| angry | 89.27% | 94.53% | 94.39% | **95.24%** |
| disgusted | 92.57% | 94.46% | **96.79%** | 95.96% |
| fearful | 87.49% | 89.63% | 89.19% | **90.67%** |
| happy | 91.57% | 95.31% | **96.71%** | 95.46% |
| neutral | 94.07% | **95.73%** | 94.94% | 95.56% |
| other | 91.62% | 95.28% | 95.64% | **95.87%** |
| sad | 93.84% | 91.76% | 92.91% | **97.34%** |
| surprised | 92.30% | 94.34% | **96.91%** | 95.27% |
| **Total** | **91.99%** | **94.20%** | **95.16%** | **95.57%** |

## 결과 3: 단계별 개선

- **Step 1** (Baseline -> 8d Skip+KL, +2.21%p): Skip + Deep MLP + KL loss로 emotion 의미 직접 학습. angry +5.26p
- **Step 2** (8d -> 9d, +0.96%p): 차원 확장 (코드북 5배), 비트 +12.5%로 한계 효율
- **Step 3** (Fine-tune, +0.41%p): 작은 lr로 sad 큰 회복 (92.91% -> 97.34%, +4.43p)

### 코드북 활용도

| 항목 | 값 |
|---|---|
| Unique 코드 | 42,524 / 1,953,125 |
| Codebook coverage | 2.18% (이론적 최대 2.47%의 88%) |
| Data diversity | 88.32% |

### 차원별 entropy (균등 분포 = log5 ≈ 2.32)

모든 9차원 entropy $2.25 \~ 2.30$ (균등도 $97 \~ 99\%$) $\rightarrow$ Codebook collapse 없음

### 차원별 Ablation

각 차원을 0.0으로 고정 시 분류 일치율:

| dim | 0 | 1 | 2 | 3 | 4 | 5 | 6 | 7 | 8 |
|---|---|---|---|---|---|---|---|---|---|
| 일치율 | 84.55% | 85.80% | 86.12% | 87.38% | 85.09% | 89.95% | **80.51%** | 91.05% | **81.11%** |
| 손실 | -11.02 | -9.77 | -9.45 | -8.19 | -10.48 | -5.62 | **-15.06** | -4.52 | **-14.46** |

9차원 모두 -4.5%p 이상 손실 $\rightarrow$ **잉여 차원 없음, 9d 차원 확장 정당화**

## 한계

- **fearful 90.67%**: 데이터 부족 (2,294개) + 음향 다양성으로 가장 약함
- **이론적 한계**: 18\~21 bits 압축에서 95\~97%가 현실적 상한. 99% 도달은 비트 30+ 필요로 압축 의미 약화
- **Fine-tune의 한계**: 50 epoch 추가 학습으로 +0.41%p에 그침

## 관련 코드 및 모델

- 코드: `src/fsq_ae/`
- 최종 모델: `data/models/fsq_ae_final.pt` (95.57%)
- 비교 모델: `fsq_ae_baseline.pt`, `fsq_ae_8d_skip_kl.pt`, `fsq_ae_9d_skip_kl.pt`

**Comment by TaeWonM** [comment]
# 추가적인 개선 방안

## 개선 방안의 이유 

가정 : Level이 높아지면, 복원률이 좋아질것이라는 가정

Level에 따라서 보내는 bit 수가 달라질텐데 2의 제곱으로 bit 수가 달라진다고 가정, 5에서 7로 바꾸든지 3bit로 일정하다는 가정

또한, 아마 홀수로 하는 것이 좋을 것으로 예상되므로 8이 아닌 8보다 작은 홀수 중 가장 큰 홀수인 7로 바꾸는 것으로 추천

**Comment by zxxxv** [comment]
# FSQ-AE Mixed Levels Ablation 결과 (추가 개선)

## 실험 개요

5L×9D 베이스라인(95.57%) 대비 7개 구성 비교. 동일 방식(200ep main + 50ep fine-tune, lr 3e-4 -> 1e-4)

## 결과 요약

| Config | Bits | Codebook | Match% | Cos% |
|--------|------|----------|--------|-----|
| 5L×9D (baseline) | 20.90 | 1.95M | 95.57 | 98.86 |
| [8,5×8] 9D | 21.58 | 3.13M | 95.53 | 99.03 |
| [8,8,5×7] 9D | 22.25 | 5.00M | 95.92 | 99.10 |
| [8,8,8,5×6] 9D | 22.92 | 8.00M | 96.26 | 99.17 |
| [6×9] 9D | 23.26 | 10.08M | 96.15 | 99.19 |
| **[8×8] 8D** | **24.00** | **16.78M** | **96.97** | **99.42** |
| [8×9] 9D | 27.00 | 134.22M | 97.18 | 99.55 |

1. **24비트 sweet spot**: 비트당 emotion match 향상 효율이 24비트에서 급감
`22.92 -> 24.00 (+1.08 bit) `: `+0.71%p -> 비트당 0.66%p`   `peak`
`24.00 -> 27.00 (+3.00 bit) `: `+0.21%p -> 비트당 0.07%p`  `saturate`
   24비트 이후 비트당 효율이 9배 이상 떨어짐. **24비트가 emotion2vec 임베딩의 본질적 정보량 한계**.

2. **9D Uniform의 dim 1 함몰**: `[6×9] 9D`, `[8×9] 9D` 모두 dim 1 entropy 83\~84% (다른 차원 92\~99%). `[8×8] 8D`만 모든 차원 98%+ 균등

3. **`[8×9] 9D`는 sparse 위험**: Coverage 0.04%, reuse 1.01 (sample마다 거의 unique 코드 = lookup-table 동작)

4. **모든 차원 필수**: 모든 모델에서 차원별 ablation 손실 -4%p 이상 -> 잉여 차원 없음

## 채택: [8×8] 8D

- 96.97% emotion preservation (+1.40%p vs baseline)
- 24비트 sweet spot
- 모든 차원 균등 활용
- 데이터 한계까지 도달 (47,386/48,148)
- `[8×9] 9D` 대비 8× 작은 코드북에 +0.21%p만 손해

## 첨부 그림

### 1. **`fsq_ablation_main_ft.png`** — Bits vs Match + Cosine (Main vs FT 점선 연결)
<img width="1929" height="1030" alt="Image" src="https://github.com/user-attachments/assets/75020dca-d7b2-48c8-b4dc-6d6b9ad8a6fc" />


### 2. **`fsq_emotion_heatmap.png`** — 감정별 분류 일치율
<img width="1665" height="883" alt="Image" src="https://github.com/user-attachments/assets/f4e22b55-cf3e-4bae-9b53-c55c74923829" />


### 3. **`fsq_codebook_analysis.png`** — 코드북 활용도 (Active / Coverage / Diversity)
<img width="2384" height="731" alt="Image" src="https://github.com/user-attachments/assets/2b4d6fb7-c805-49cb-b7d0-896471bd69ec" />


### Per-Dimension Ablation (각 차원 0으로 ablate 시 분류 일치율 손실)

| Config | base | dim 0 | dim 1 | dim 2 | dim 3 | dim 4 | dim 5 | dim 6 | dim 7 | dim 8 | min | max |
|--------|------|-------|-------|-------|-------|-------|-------|-------|-------|-------|-----|-----|
| [8,5×8] 9D | 95.53 | -9.18 | -10.30 | -6.79 | -10.94 | -11.46 | -9.38 | -9.34 | -5.62 | -13.90 | -5.62 | -13.90 |
| [8,8,5×7] 9D | 95.92 | -16.52 | -4.62 | -5.85 | -11.88 | -9.43 | -8.90 | -8.21 | -4.06 | -15.06 | -4.06 | -16.52 |
| [8,8,8,5×6] 9D | 96.26 | -11.88 | -6.79 | -17.76 | -4.37 | -14.67 | -6.98 | -10.37 | -12.51 | -7.74 | -4.37 | -17.76 |
| [6×9] 9D | 96.15 | -8.00 | -6.35 | -12.44 | -9.77 | -11.53 | -12.13 | -13.68 | -7.73 | -8.76 | -6.35 | -13.68 |
| [8×8] 8D | 96.97 | -11.70 | -10.81 | -15.37 | -13.41 | -10.97 | -13.76 | -10.03 | -7.51 | — | -7.51 | -15.37 |
| [8×9] 9D | 97.18 | -8.80 | -9.39 | -11.44 | -10.24 | -11.75 | -12.07 | -15.11 | -8.64 | -7.18 | -7.18 | -15.11 |

**모든 차원 -4%p 이상 손실 -> 잉여 차원 없음**

**Comment by zxxxv** [comment]
# [FSQ-AE] KL Weight Ablation — λ=0.5 채택

## 결론
**KL weight λ = 0.5 채택**

- 8d × 8L (24bit) 구성, 200 epoch main + 50 epoch fine-tune (lr=1e-4)
- 성능: match **97.64%**, val_cos 0.9933, val_mse 0.0118

## 결과

| KL  | Match% | Cos    | MSE     |
|-----|--------|--------|---------|
| 0.1 | 96.97  | 0.9942 | 0.00972 |
| 0.3 | 97.50  | 0.9939 | 0.01034 |
| **0.5** | **97.64** | 0.9933 | 0.01178 |
| 0.7 | 97.56  | 0.9934 | 0.01124 |
| 1.0 | 97.76  | 0.9930 | 0.01200 |

## 채택 이유
- KL=1.0이 match% 최고지만 0.5와 +0.12%p 차이
- KL=1.0은 angry 감정 보존 -1.5%p 하락 (편향 발생)
- KL=0.5는 학습 안정성 우수 (Main 200ep에서 자연 수렴)
- 분류 정확도 + 임베딩 quality + 감정 균형 모두 양호

### 상황별 추천

| 상황 / 우선순위 | 추천 |
|---|---|
| 임베딩 quality / Reconstruction 우선 | **0.1** |
| 모든 차원 균등 활용 (정보 압축 효율) | **0.1** |
| 균형 + MSE 살짝 우선 | **0.3** |
| **균형 (가장 무난, 현재 채택)** | **0.5** |
| 균형 + 분포 보존 우선 | **0.7** |
| 분류 정확도 절대 우선 | **1.0** |
| 감정 편향 최소화 (angry/sad 보존) | **0.5** |
| 학습 안정성 / 재현성 | **0.3 ~ 0.5** |
